[DO NOT MERGE] Staging area for upstream PR

mbrookhart opened the pull request.

Commit:
MXNet - nGraph initial integration

ashokei commented on Makefile:
```diff
@@ -71,11 +80,14 @@ ifeq ($(USE_MKLDNN), 1)
+
 # all tge possible warning tread
-WARNFLAGS= -Wall -Wsign-compare
+WARNFLAGS= -Wall -Wsign-compare -Wno-comment
```
can we try to limit our changes to ngraph submodule , unless we need to

mbrookhart replied:
when we include ngraph.hpp in ndarray.h, it prints a lot of warnings from doxygen code comments for anything that includes ndarray.h

mbrookhart replied:
This one, in particular, is needed because we include ngraph.hpp in ndarray.h, and gcc really doesn't like the latex in the code comments in ngraph.hpp, so every file that includes ndarray.h dumps dozens of comment warnings.

ashokei commented on make/config.mk:
```diff
@@ -101,6 +102,36 @@ USE_OPENMP = 1
 # whether use MKL-DNN library
 USE_MKLDNN = 0
 
+# The following variables influence if/how MXnet is built with nGraph support:
```
this file maybe we can limit to our initial PR related changes

ashokei commented on CMakeLists.txt:
```diff
@@ -18,9 +18,10 @@ mxnet_option(USE_CUDNN            "Build with cudnn support"  ON) # one could se
 mxnet_option(USE_SSE              "Build with x86 SSE instruction support" ON IF NOT ARM)
 mxnet_option(USE_F16C             "Build with x86 F16C instruction support" ON) # autodetects support if ON
 mxnet_option(USE_LAPACK           "Build with lapack support" ON)
+mxnet_option(USE_NGRAPH           "Build with nGraph support" ON)
```
this needs to be OFF ?

ashokei commented on CMakeLists.txt:
```diff
@@ -404,7 +434,7 @@ endif()
 
 # ---[ LAPack
 if(USE_LAPACK)
-  message("USE_LAPACK is ON")
```
is this removed in upstream master?

mbrookhart replied:
Nope, will fix.

ashokei commented on Makefile:
```diff
@@ -90,14 +102,19 @@ else
 CFLAGS += -I$(TPARTYDIR)/mshadow/ -I$(TPARTYDIR)/dmlc-core/include -fPIC -I$(NNVM_PATH)/include -I$(DLPACK_PATH)/include -I$(TPARTYDIR)/tvm/include -Iinclude $(MSHADOW_CFLAGS)
-LDFLAGS = -pthread $(MSHADOW_LDFLAGS) $(DMLC_LDFLAGS)
 
+LDFLAGS =
```
are we resetting this to empty (if there were exported or set previously?)

mbrookhart replied:
No, this is the first place it's defined, master overwrites it hear as well.

ashokei commented on src/ndarray/ndarray.cc:
```diff
@@ -39,11 +43,13 @@
 #include "../operator/tensor/matrix_op-inl.h"
 #include "../operator/tensor/init_op.h"
 #include "../operator/nn/mkldnn/mkldnn_base-inl.h"
+#include "../engine/engine_impl.h"
```
may not be needed.

mbrookhart replied:
Not sure where that came from.

ashokei commented on .gitmodules:
```diff
@@ -29,3 +29,6 @@
 	url = https://github.com/onnx/onnx-tensorrt.git
+[submodule "3rdparty/ngraph-mxnet-bridge"]
+	path = 3rdparty/ngraph-mxnet-bridge
+	url = https://github.com/NervanaSystems/ngraph-mxnet-bridge
```
may want to add , so it doesnt show dirty when build	:
`ignore = untracked`

luobao-intel committed:
Fix flaky test: test_mkldnn.test_activation #12377 (#12418)

* test_activation_rec_eps

* enable case

roywei committed:
allow foreach on input with 0 length (#12471)

* allow foreach on input with 0 length

* add test foreach with unknown dim

Commit:
[MXNET-696][PYTHON][UNDEFINED NAME] import logging in ci/util.py (#12488)

Fixes #12406 @larroy @marcoabreu @szha 

[flake8](http://flake8.pycqa.org) testing of https://github.com/apache/incubator-mxnet on Python 3.7.0

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./ci/util.py:73:21: F821 undefined name 'logging'
                    logging.warning("Exception: %s, Retrying in %d seconds...", str(e), mdelay)
                    ^
1     F821 undefined name 'logging'
1
```

KellenSunderland committed:
[MXNET-703] Static linking for libprotobuf with TensorRT (#12475)
